Updated README (#841)
krzys-h/UndertaleModTool: .github/README.md

<p align="center">

>   <img src="images/logo.png" alt="UndertaleModTool Logo"/>

Reviewer: As was said before in the server, I feel like the title is a little *too* big,

</p>

![Discord](https://img.shields.io/discord/566861759210586112?label=Discord&logo=discord&logoColor=white) ![GitHub](https://img.shields.io/github/license/krzys-h/UndertaleModTool?logo=github)

<p align="center">(seeing such an amazing tool it fills you with DETERMINATION.)</p>
<p align="center">Heya. I heard you like digging deep into Undertale/Deltarune or any GameMaker game data so I made a tool just for you!</p>
<p align="center">

>   <img src="images/flowey.gif" alt="Flowey"/>

Reviewer: Something the previous image showed well that this one doesn't, is that UMT is actually modifying game data.

Author: has i say in the server the image doesn't show anything remember that before utmt you could activate the console exclusive  content found in 1.0.8 with hex editing including the borders, if we want to add screenshots we need to show something that you can't do in Hex

</p>

# Downloads

Looking for the most stable or the latest avaliable download? Both the latest stable release and the most cutting edge version are in the table below!

| Releases 	| Status 	|
|:---:	|----------	|

> | Stable[^1][^2] 	| ![GitHub release (latest by date)](https://img.shields.io/github/downloads/krzys-h/UndertaleModTool/latest/total) |

Reviewer: Clicking the release badge does not link to the actual release page which would be devastating for new users.

| Bleeding edge[^1] 	| [![Build tool NET Bundled](https://github.com/krzys-h/UndertaleModTool/actions/workflows/build_net.yml/badge.svg)](https://github.com/krzys-h/UndertaleModTool/actions/workflows/build_net.yml) |

> [^1]: UndertaleModTool has multiple releases such has .net bundle, non single file, and single file the differenes on these releases are:

Reviewer: ```suggestion
[^1]: UndertaleModTool has different builds per release. The differences are as follows:
```

>       - .NET bundle: bundles the required .net framework version to run the tool

Reviewer: ```suggestion
      - `.NET bundled` - bundles the required .NET runtime version needed to run the tool alongside it. All stable releases are .NET bundled, so you don't have to worry about installing the necessary runtime.
```

>       - Single file: just one exe an no extra dlls

Reviewer: ```suggestion
      - `Single file` - the tool is only one executable, with all dependencies embedded within it. This does make your folders cleaner, however it also causes some unexpected stability issues.
```

>       - Non Single File: Just like above but it ships the dlls that are required by the tool

Reviewer: ```suggestion
      - `Non-single File` - all dependencies are not embedded within the executable, but are now located right next to it. Choose this if you don't care about finding the right executable in-between of ~300 dll's, or if the stability issues from the `Single file` build affect you.
```

> [^2]: You can updated to the blending edge releases within the settings menu in UndertaleModTool.

Reviewer: Those footnotes are shown alllll the way at the bottom of the file.
Would be nicer to actually display them right under the download table. 
Think that not as footnotes would help more for people to actually read the notes, as it would make them less missable.

Author: that's the way of how markdown handles footnotes, there's no much i can do about it

Reviewer: There is a way, I'll make a new comment for this in a few to show potential solutions.

> [^2]: You can updated to the blending edge releases within the settings menu in UndertaleModTool.

Reviewer: ```suggestion
[^2]: You can update to the bleeding edge releases at any time from within the settings menu of UndertaleModTool.
```

> | Releases 	| Status 	|
> |:---:	|----------	|
> | Stable[^1][^2] 	| [![GitHub release (latest by date)](https://img.shields.io/github/downloads/krzys-h/UndertaleModTool/latest/total)](https://github.com/krzys-h/UndertaleModTool/releases/latest) |
> | Bleeding edge[^1] 	| [![Build tool NET Bundled](https://github.com/krzys-h/UndertaleModTool/actions/workflows/build_net.yml/badge.svg)](https://github.com/krzys-h/UndertaleModTool/actions/workflows/build_net.yml) |
> 
> [^1]: UndertaleModTool has different builds per release. The differences are as follows:
>       - `.NET bundled` - bundles the required .NET runtime version needed to run the tool alongside it. All stable releases are .NET bundled, so you don't have to worry about installing the necessary runtime.
>       - `Single file` - the tool is only one executable, with all dependencies embedded within it. This does make your folders cleaner, however it also causes some unexpected stability issues.
>       - `Non-single File` - all dependencies are not embedded within the executable, but are now located right next to it. Choose this if you don't care about finding the right executable in-between of ~300 dll's, or if the stability issues from the `Single file` build affect you.
> [^2]: You can update to the bleeding edge releases at any time from within the settings menu of UndertaleModTool.
> 

Reviewer: Option 1 use the in-built superscript notations for footnotes:
```suggestion
| Releases 	| Status 	|
|:---:	|----------	|
| Stable¹ ² 	| [![GitHub release (latest by date)](https://img.shields.io/github/downloads/krzys-h/UndertaleModTool/latest/total)](https://github.com/krzys-h/UndertaleModTool/releases/latest) |
| Bleeding edge¹ 	| [![Build tool NET Bundled](https://github.com/krzys-h/UndertaleModTool/actions/workflows/build_net.yml/badge.svg)](https://github.com/krzys-h/UndertaleModTool/actions/workflows/build_net.yml) |

¹: UndertaleModTool has different builds per release. The differences are as follows:
      - `.NET bundled` - bundles the required .NET runtime version needed to run the tool alongside it. All stable releases are .NET bundled, so you don't have to worry about installing the necessary runtime.
      - `Single file` - the tool is only one executable, with all dependencies embedded within it. This does make your folders cleaner, however it also causes some unexpected stability issues.
      - `Non-single File` - all dependencies are not embedded within the executable, but are now located right next to it. Choose this if you don't care about finding the right executable in-between of ~300 dll's, or if the stability issues from the `Single file` build affect you.
²: You can update to the bleeding edge releases at any time from within the settings menu of UndertaleModTool.

```

Option 2 get rid of footnotes entirely:
```suggestion
| Releases 	| Status 	|
|:---:	|----------	|
| Stable 	| [![GitHub release (latest by date)](https://img.shields.io/github/downloads/krzys-h/UndertaleModTool/latest/total)](https://github.com/krzys-h/UndertaleModTool/releases/latest) |
| Bleeding edge 	| [![Build tool NET Bundled](https://github.com/krzys-h/UndertaleModTool/actions/workflows/build_net.yml/badge.svg)](https://github.com/krzys-h/UndertaleModTool/actions/workflows/build_net.yml) |

It is worth noting that UndertaleModTool has different builds per release. The differences are as follows:
      - `.NET bundled` - bundles the required .NET runtime version needed to run the tool alongside it. All stable releases are .NET bundled, so you don't have to worry about installing the necessary runtime.
      - `Single file` - the tool is only one executable, with all dependencies embedded within it. This does make your folders cleaner, however it also causes some unexpected stability issues.
      - `Non-single File` - all dependencies are not embedded within the executable, but are now located right next to it. Choose this if you don't care about finding the right executable in-between of ~300 dll's, or if the stability issues from the `Single file` build affect you.
You can also update to the bleeding edge releases at any time from within the settings menu of UndertaleModTool.

```

I personally prefer option 2 more, as it reads more like a proper text and the flow of reading is not interrupted with the footnotes.

# Downloads

> Looking for the most stable or the latest avaliable download? Both the latest stable release and the most cutting edge version are in the table below!

Reviewer: Also just noticed it:
`avaliable` -> `available`
Personally think this should get a little rewritten due to the first sentence essentially saying the same thing as the second. Maybe more like:
`Looking for the latest stable or the most cutting edge download? Look in the table below!` ? Or just ditch the first sentence entirely?
Maybe just `Both the latest stable release and the most cutting edge version can be downloaded from the table below!` ?

# Downloads

> oth the latest stable release and the most cutting edge version can be downloaded from the table below!

Reviewer: Another typo

* All core IO functionality extracted into a library for use in external tools.
* Can generate an .yydebug file for the GM:S debugger so that you can edit variables live! (see [here](https://github.com/krzys-h/UndertaleModTool/wiki/Corrections-to-GameMaker-Studio-1.4-data.win-format-and-VM-bytecode,-.yydebug-format-and-debugger-instructions#yydebug-file-format))

> # Include Scripts

Reviewer: ```suggestion
# Included Scripts
```

Included are some test scripts. They are, but not limited to:

* Universal:

>   * EnableDebug: does just that, makes the global variable 'debug' be enabled at game start. If you don't know about Undertale's debug mode, check out [TCR Undertale Debug Mode](https://tcrf.net/Undertale/Debug_Mode)

Reviewer: ```suggestion
  * EnableDebug: does just that, makes the global variable 'debug' be enabled at game start. If you don't know about Undertale's debug mode, check out the [this write up of Undertale's Debug Mode](https://tcrf.net/Undertale/Debug_Mode)
```

* All core IO functionality extracted into a library for use in external tools.
* Can generate an .yydebug file for the GM:S debugger so that you can edit variables live! (see [here](https://github.com/krzys-h/UndertaleModTool/wiki/Corrections-to-GameMaker-Studio-1.4-data.win-format-and-VM-bytecode,-.yydebug-format-and-debugger-instructions#yydebug-file-format))

# Included Scripts

Included are some test scripts. They are, but not limited to:

* Universal:

>   * EnableDebug: does just that, makes the global variable 'debug' be enabled at game start. If you don't know about Undertale's debug mode, check out the [this write up of Undertale's Debug Mode](https://tcrf.net/Undertale/Debug_Mode)

Reviewer: probably a typo from me
`the this`. the `the` should get removed.

* All core IO functionality extracted into a library for use in external tools.
* Can generate an .yydebug file for the GM:S debugger so that you can edit variables live! (see [here](https://github.com/krzys-h/UndertaleModTool/wiki/Corrections-to-GameMaker-Studio-1.4-data.win-format-and-VM-bytecode,-.yydebug-format-and-debugger-instructions#yydebug-file-format))

# Screenshots

Here's some screenshots of what UTMT can do:

## [AM2R: Horde Mode Mod](https://github.com/Hornoads/AM2R-The-Horde-Multitroid/releases)
<img src="images/am2r.png" alt="AM2R: Horde Mode"/>

## [Undertale: Bits and Pieces Mod](https://gamejolt.com/games/UndertaleBnP/574044)
<img src="images/bnp-ut.png" alt="BnP"/>

## [RIBBIT - The Deltarune Mod](https://gamejolt.com/games/ribbitmod/671888)

> <img src="images/ribbit-dr.png" alt="RIBBIT"/>

Reviewer: Can you make these images have a consistent size please?

Author: that would lead to some image artifacts

Reviewer: Wouldn't be very noticeable, due to the small size anyway. I just don't feel it looks good if
a) all images have varying sizes
b) the readme is cluttered with huge images.

  * DeltaHATE: [HATE](https://www.reddit.com/r/Undertale/comments/41lb16/hate_the_undertale_corruptor/)-inspired script for corrupting Deltarune
  * DeltaMILK: Replaces every non-background sprite with the K.Round healing milk. Don't ask why.
  * TheWholeWorldRevolving: The world is spinning, spinning

Additionally, included are some community-made scripts. For more information, consult the [SCRIPTS.md](https://github.com/krzys-h/UndertaleModTool/blob/master/SCRIPTS.md) file.

> # Bug reports, Contributing

Reviewer: I feel like just `# Contributing` is clearer.

Alternatively, you can also compile via command line, like so: `dotnet publish UndertaleModTool --no-self-contained -p:PublishSingleFile=true -c release -r win-x86`
You can adjust win-x86 to other RIDs, such as win-x64 or win-arm.

> 

Reviewer: Technically not part of this PR, but the compilation instructions should get rewritten. I may do it at one point later, but the key aspects are:
- Windows is only needed for compiling the GUI
- due to all projects now having rolling-forward enabled, .NET core 5 is not needed anymore, rather anything above it
- Visual Studio is not needed; only the SDK. Visual studio just invokes the command line parameters, which can be done by anything (visual studio code, rider, vim, etc.)
- the Cli parameters are technically not necessary
- no description of what RIDs are.

Reviewer: This wasn't addressed in any way, why was this closed?

Reviewer: Rewrote the part.
```md
In order to compile the repo yourself, the `.NET Core 5 SDK` or later is required.

The following projects can be compiled:  
- `UndertaleModLib`: The core library used by all other projects.
- `UndertaleModCli`: A command line interface for interacting with GameMaker data files and applying scripts. Currently is very primitive in what it can do.
- `UndertaleModTool`: The main graphical user interface for interacting with GameMaker data files. **Windows is required in order to compile this**.

#### Compiling Via IDE
- Open the `UndertaleModTool.sln` in the IDE of your choice (Visual Studio, JetBrains Rider, Visual Studio Code etc.)
- Select the project you want to compile
- Compile

#### Compiling Via Command Line
- Open a terminal and navigate to the directory of `UndertaleModTool.sln`
- Execute `dotnet publish <Project>` where `<Project>` is one of the projects listed above.
You can also provide arguments for compiling, such as `--no-self-contained` or `-c release`. For a full list of arguments, consult the [Microsoft Documentation](https://docs.microsoft.com/dotnet/core/tools/dotnet-publish)
```

> # data.win File Format

Reviewer: considering the `data.win` can have different names, i'd personally rename this to something more akin to `GameMaker Data File Format`

Interested in the file and instruction format research I've done while working on this? Check out these:

* [Corrections to GameMaker Studio 1.4 data.win format and VM bytecode, .yydebug format and debugger instructions](https://github.com/krzys-h/UndertaleModTool/wiki/Corrections-to-GameMaker-Studio-1.4-data.win-format-and-VM-bytecode,-.yydebug-format-and-debugger-instructions)
* [Changes in GameMaker Studio 2](https://github.com/krzys-h/UndertaleModTool/wiki/Changes-in-GameMaker-Studio-2)
* [Extensions, Shaders, Timelines format](https://github.com/krzys-h/UndertaleModTool/wiki/Extensions,-Shaders,-Timelines-format)
* [Bytecode version differences](https://github.com/krzys-h/UndertaleModTool/wiki/Bytecode-version-differences)
* [YYC games](https://github.com/krzys-h/UndertaleModTool/wiki/YYC-games)

> feel free to check out the [Wiki](https://github.com/krzys-h/UndertaleModTool/wiki) for more.

Reviewer: *All* of the links are on the wiki, so why not just link directly to the wiki? Would clutter the readme less, and still get the point across.

Reviewer 2: Also why does the YYC games page still exist, it's been empty for like 4 years now.

# Special thanks

Special thanks to everybody who did previous research on unpacking and decompiling Undertale, it was a really huge help:

> * [Ulyssis's UNDERTALE decompilation research](https://pcy.ulyssis.be/undertale/)

Reviewer: ```suggestion
* [Ulyssis' UNDERTALE decompilation research](https://pcy.ulyssis.be/undertale/)
```

Reviewer 3: But mirrawrs is the one who made the original site, right? There's only a few changes on the rehosted version

Reviewer: So is the research from mirrawrs or ulyssis?

* [Donkeybonks's GameMaker data.win Bytecode research](https://web.archive.org/web/20191126144953if_/https://github.com/donkeybonks/acolyte/wiki/Bytecode)
* [PoroCYon's Altar.NET](https://github.com/PoroCYon/Altar.NET)
* [WarlockD's GMdsam](https://github.com/WarlockD/GMdsam)
* [@NarryG](https://github.com/NarryG) for [helping me figure out](https://github.com/krzys-h/UndertaleModTool/issues/3) the missing stuff for GMS2 and Nintendo Switch release
* [@colinator27](https://github.com/colinator27) for [lots of things, including the gml compiler](https://github.com/krzys-h/UndertaleModTool/issues/4), [Sha](https://github.com/krzys-h/UndertaleModTool/issues/13)[ders](https://github.com/krzys-h/UndertaleModTool/pull/25) and [a bunch of other stuff](https://github.com/krzys-h/UndertaleModTool/pull/30)

> * [@Kneesnap](https://github.com/Kneesnap) for [improving the decompiler a bunch](https://github.com/krzys-h/UndertaleModTool/pull/162)

Reviewer: I'd personally omit these from the special thanks, as they're already covered by the "all contributors" list down below.

as well as all the other contributors:
<p align="center">
  <a href="https://github.com/krzys-h/UndertaleModTool/graphs/contributors">

>     <img src="https://contrib.rocks/image?repo=krzys-h/UndertaleModTool" />

Reviewer: Wouldn't this massively clutter the readme, if at one point UMT gets >100 contributors or so?

Reviewer 3: That implies we can't change it before that happens... while we have a smaller amount of contributors it's fine as-is

Reviewer: Was not the intent to imply that, more just a general question / thought food on how it should be handled in the future whenever it gets to that point.

Author: this list imaged is limited don't worry about it

  </a>
</p>

And of course, special thanks to Toby Fox and the whole Undertale team for making the game(s) ;)

<p align="center">
  <img src="images/papyrus.gif" alt="Papyrus"/>

> </p>

Reviewer: Unnecessary filler which doesn't convey any meaning.

Author: this was added because was the original comment that krzys added into the readme

Reviewer: well tbh the original comment didn't have much value either.
